### ISSUE TITLE
[LSP] shutdown and exit commands throw an exception

### ISSUE THREAD

**Issue by LoneBoco** [OmniSharp/omnisharp-roslyn#1113]
I am trying to use omnisharp-roslyn in LSP mode and I am unable to shut down the server through normal means.

```
LSP:  --> shutdown
server: [dbug]: OmniSharp.Extensions.LanguageServer.LspRequestRouter
        Finding descriptor for shutdown
LSP:server: [crit]: OmniSharp.Extensions.LanguageServer.LspRequestRouter
        Failed to handle notification shutdown
System.Reflection.TargetParameterCountException: Parameter count mismatch.
   at System.Reflection.RuntimeMethodInfo.InvokeArgumentsCheck(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at OmniSharp.Extensions.JsonRpc.ReflectionRequestHandlers.HandleRequest(IHandlerDescriptor instance, CancellationToken token)
   at OmniSharp.Extensions.LanguageServer.LspRequestRouter.<RouteRequest>d__11.MoveNext()
```

I tried throwing /shutdownServer to see what would happen (it is currently listed as not supported), and I get this response:

```
LSP:  --> shutdownServer
server: [dbug]: OmniSharp.Extensions.LanguageServer.LspRequestRouter
        Finding descriptor for shutdownServer
server: [dbug]: OmniSharp.Extensions.LanguageServer.LspRequestRouter
        Unable to find shutdownServer, methods found include initialize:OmniSharp.Extensions.LanguageServer.LanguageServer, initialized:OmniSharp.Extensions.LanguageServer.LanguageServer, shutdown:OmniSharp.Extensions.LanguageServer.Handlers.ShutdownHandler, exit:OmniSharp.Extensions.LanguageServer.Handlers.ExitHandler, $/cancelRequest:OmniSharp.Extensions.LanguageServer.Handlers.CancelRequestHandler, textDocument/didChange:OmniSharp.LanguageServerProtocol.Handlers.TextDocumentSyncHandler, textDocument/didOpen:OmniSharp.LanguageServerProtocol.Handlers.TextDocumentSyncHandler, textDocument/didClose:OmniSharp.LanguageServerProtocol.Handlers.TextDocumentSyncHandler, textDocument/didSave:OmniSharp.LanguageServerProtocol.Handlers.TextDocumentSyncHandler, textDocument/willSave:OmniSharp.LanguageServerProtocol.Handlers.TextDocumentSyncHandler, textDocument/willSaveWaitUntil:OmniSharp.LanguageServerProtocol.Handlers.TextDocumentSyncHandler, textDocument/didChange:OmniSharp.LanguageServerProtocol.Handlers.TextDocumentSyncHandler, textDocument/didOpen:OmniSharp.LanguageServerProtocol.Handlers.TextDocumentSyncHandler, textDocument/didClose:OmniSharp.LanguageServerProtocol.Handlers.TextDocumentSyncHandler, textDocument/didSave:OmniSharp.LanguageServerProtocol.Handlers.TextDocumentSyncHandler, textDocument/willSave:OmniSharp.LanguageServerProtocol.Handlers.TextDocumentSyncHandler, textDocument/willSaveWaitUntil:OmniSharp.LanguageServerProtocol.Handlers.TextDocumentSyncHandler, textDocument/definition:OmniSharp.LanguageServerProtocol.Handlers.DefinitionHandler, textDocument/definition:OmniSharp.LanguageServerProtocol.Handlers.DefinitionHandler, textDocument/hover:OmniSharp.LanguageServerProtocol.Handlers.HoverHandler, textDocument/hover:OmniSharp.LanguageServerProtocol.Handlers.HoverHandler, textDocument/completion:OmniSharp.LanguageServerProtocol.Handlers.CompletionHandler, textDocument/completion:OmniSharp.LanguageServerProtocol.Handlers.CompletionHandler, textDocument/signatureHelp:OmniSharp.LanguageServerProtocol.Handlers.SignatureHelpHandler, textDocument/signatureHelp:OmniSharp.LanguageServerProtocol.Handlers.SignatureHelpHandler, textDocument/rename:OmniSharp.LanguageServerProtocol.Handlers.RenameHandler, textDocument/documentSymbol:OmniSharp.LanguageServerProtocol.Handlers.DocumentSymbolHandler, textDocument/documentSymbol:OmniSharp.LanguageServerProtocol.Handlers.DocumentSymbolHandler
```

shutdown and exit seem to be handled somewhere else.  I tried briefly searching through the code of omnisharp-roslyn, but I couldn't find out where it was handled.

```
shutdown:OmniSharp.Extensions.LanguageServer.Handlers.ShutdownHandler
exit:OmniSharp.Extensions.LanguageServer.Handlers.ExitHandler
```

EDIT: I should mention that I am running Omnisharp with the -lsp and -stdio command line arguments.

**Comment by david-driscoll** [comment]
This should be fixed in the next version of the lsp library (that we have not yet updated to).

@mholo65 is working on a code lens handler, and the upgrade will roll in with that.

**Comment by LoneBoco** [comment]
Thanks for that hint.  I was able to figure out how it works and merge the LSP library version 0.7.7 into my own code.  I have discovered a bunch of other problems, but I don't know where they should be reported.  I guess in csharp-language-server-protocol?

**Comment by LoneBoco** [comment]
Well, the bug with OmniSharp.exe not terminating is because @mholo65 's current code doesn't bind the server's Exit event in LanguageServerHost.cs.  Since you said it was a work in progress, I'll just leave it alone for now.

**Comment by david-driscoll** [comment]
@LoneBoco the language server library actually manages exit and shutdown for you @ https://github.com/OmniSharp/csharp-language-server-protocol/blob/master/src/Server/LanguageServer.cs#L88

**Comment by david-driscoll** [comment]
However we did have an issue with shutdown and exit not working correctly, that has been fixed in the library, it just hasn't been merged yet.

**Comment by LoneBoco** [comment]
Is that so?  Does the library forcibly kill the process when that happens?

According to this, Omnisharp waits on the cancellation token when we start it up:
https://github.com/OmniSharp/omnisharp-roslyn/blob/7da2d946d9c698d771e874f70ee0d42e09e7c7f9/src/OmniSharp.Stdio/Program.cs#L41

And here we can see that we don't send the cancellation token unless we abort in the console, or unless the host process terminates:
https://github.com/OmniSharp/omnisharp-roslyn/blob/7da2d946d9c698d771e874f70ee0d42e09e7c7f9/src/OmniSharp.LanguageServerProtocol/LanguageServerHost.cs#L198-L218

There is nothing that actually triggers that cancellation token if we receive an exit command.

For what it is worth, I added a simple block of code and it now exits for me:

```c#
_server.Exit += (sender) =>
{
    _cancellationTokenSource.Cancel();
};
```

This is because the library sends out that Exit event when it receives the exit command.

**Comment by mholo65** [comment]
Available in 1.32.10
